trillian/migrillian/election/etcd/election.go

```diff
@@ -0,0 +1,152 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package etcd provides implementation master election based on etcd.
+package etcd
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+	"github.com/golang/glog"
+	"github.com/google/certificate-transparency-go/trillian/migrillian/election"
+)
+
+// Election is an implementation of election.Election based on etcd.
+type Election struct {
+	treeID     int64
+	instanceID string
+	lockFile   string
+
+	client   *clientv3.Client
+	session  *concurrency.Session
+	election *concurrency.Election
+
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+// Await blocks until the instance captures mastership. Returns the context
+// which remains active until the instance stops believing to be the master,
+// or the passed in context is canceled. Returns an error if capturing fails
+// or the passed in context is canceled before that.
+func (e *Election) Await(ctx context.Context) (context.Context, error) {
+	if e.ctx != nil && e.ctx.Err() == nil {
+		return e.ctx, nil
+	}
+
+	if err := e.election.Campaign(ctx, e.instanceID); err != nil {
+		return nil, err
+	}
+
+	cctx, cancel := context.WithCancel(ctx)
+	ch := e.election.Observe(cctx)
+
+	select {
+	case <-ctx.Done():
+		cancel()
+		return nil, ctx.Err()
+	case rsp, ok := <-ch:
+		if !ok {
+			cancel()
+			return nil, errors.New("mastership unconfirmed")
+		}
+		if string(rsp.Kvs[0].Value) != e.instanceID {
+			cancel()
+			return nil, errors.New("mastership overtaken")
+		}
+	}
+
+	go func() {
+		for rsp := range ch {
+			if string(rsp.Kvs[0].Value) != e.instanceID {
+				break
+			}
+		}
+		glog.Warningf("%d: canceling master context", e.treeID)
+		cancel()
+	}()
+
+	e.ctx, e.cancel = cctx, cancel
+	return cctx, nil
+}
+
+// Resign releases mastership for this instance. The instance is still ready
+// to be elected again using the Await method.
+func (e *Election) Resign(ctx context.Context) error {
+	if e.cancel != nil {
+		e.cancel()
+	}
+	return e.election.Resign(ctx)
```

Comment:
(Thinking aloud) What would happen if you resigned before cancelling?

Reply:
Well, it would succeed (unless an error, or master context is canceled due to overtake happening concurrently). I will do the shuffle (also in `Close`).

Reply:
Done.